File: packages/workspace_flutter/android/src/main/kotlin/com/mattsp1290/workspace_flutter/internal/WorkspaceOperationRegistry.kt
package com.mattsp1290.workspace_flutter.internal

import android.os.CancellationSignal
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/** The first transition is authoritative; platform cancellation stays cooperative. */
internal data class WorkspaceTerminalState(val errorCode: String?)

/** Engine-owned cancellation and cleanup identity for one native operation. */
internal class WorkspaceNativeOperation(
  val workspaceId: String,
  val engineGeneration: Long,
  val cancellation: CancellationSignal = CancellationSignal(),
  private val requestPlatformCancellation: () -> Unit = { cancellation.cancel() },
) {
  val settled = AtomicBoolean(false)
  private val cancellationRequested = AtomicBoolean(false)
  private val terminal = AtomicReference<WorkspaceTerminalState?>(null)

  fun cancel() {
    claimTerminal("cancelled")
    if (cancellationRequested.compareAndSet(false, true)) requestPlatformCancellation()
  }

  fun close() {
    claimTerminal("closed")
    if (cancellationRequested.compareAndSet(false, true)) requestPlatformCancellation()
  }

  /** Captures a provider result only when no earlier terminal event won. */
  fun captureBody(errorCode: String?): WorkspaceTerminalState? {
    val captured = WorkspaceTerminalState(errorCode)
    return if (terminal.compareAndSet(null, captured)) captured else null
  }

  fun terminalState(): WorkspaceTerminalState? = terminal.get()

  private fun claimTerminal(errorCode: String) {
    terminal.compareAndSet(null, WorkspaceTerminalState(errorCode))
  }

  /** The engine only exposes this signal to providers; it owns cancellation. */
  fun isCancelled(): Boolean = cancellationRequested.get()
}

/**
 * Thread-safe active-operation registry. Cleanup remains owned by the caller
 * until [remove] succeeds, which is what makes workspace close barriers
 * independent of MethodChannel result delivery.
 */
internal class WorkspaceOperationRegistry(
  private val operationFactory: (String, Long) -> WorkspaceNativeOperation =
    { workspaceId, generation -> WorkspaceNativeOperation(workspaceId, generation) },
) {
  private val active = ConcurrentHashMap<String, WorkspaceNativeOperation>()
  @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
  private val cleanupMonitor = java.lang.Object()
  private val closingWorkspaces = ConcurrentHashMap.newKeySet<String>()

  fun register(
    operationId: String,
    workspaceId: String,
    engineGeneration: Long,
  ): WorkspaceNativeOperation? {
    val operation = operationFactory(workspaceId, engineGeneration)
    synchronized(cleanupMonitor) {
      if (workspaceId in closingWorkspaces) return null
      return if (active.putIfAbsent(operationId, operation) == null) operation else null
    }
  }

  fun remove(operationId: String, operation: WorkspaceNativeOperation) {
    synchronized(cleanupMonitor) {
      active.remove(operationId, operation)
      cleanupMonitor.notifyAll()
    }
  }

  fun cancel(operationId: String) {
    active[operationId]?.cancel()
  }

  /** Prevents new work before requesting cancellation of existing work. */
  fun closeWorkspace(workspaceId: String) {
    closingWorkspaces.add(workspaceId)
    active.values.filter { it.workspaceId == workspaceId }.forEach { it.close() }
  }

  fun isWorkspaceClosing(workspaceId: String): Boolean = workspaceId in closingWorkspaces

  fun cancelAll() {
    active.values.forEach { it.cancel() }
  }

  fun hasLiveWorkspace(workspaceId: String): Boolean =
    active.values.any { it.workspaceId == workspaceId }

  /** Blocks only a background cleanup owner until matching work is released. */
  fun awaitWorkspaceCleanup(workspaceId: String) {
    synchronized(cleanupMonitor) {
      while (hasLiveWorkspace(workspaceId)) cleanupMonitor.wait()
    }
  }

  val activeCount: Int get() = active.size
}

File: packages/workspace_flutter/ios/Classes/Internal/WorkspaceOperationRegistry.swift
import Foundation

/// Engine-owned operation state. The coordinator stays alive until the caller
/// removes this operation after provider cleanup, not merely after a Flutter
/// result has been queued.
internal final class WorkspaceNativeOperation {
  let workspaceId: String
  let engineGeneration: Int
  let coordinator = NSFileCoordinator()
  private let lock = NSLock()
  private var cancelled = false

  init(workspaceId: String, engineGeneration: Int) {
    self.workspaceId = workspaceId
    self.engineGeneration = engineGeneration
  }

  func cancel() {
    lock.lock()
    cancelled = true
    lock.unlock()
    coordinator.cancel()
  }

  func isCancelled() -> Bool {
    lock.lock()
    defer { lock.unlock() }
    return cancelled
  }
}

/** Thread-safe active-operation registry used by the production plugin and XCTest. */
internal final class WorkspaceOperationRegistry {
  private let cleanupCondition = NSCondition()
  private var active: [String: WorkspaceNativeOperation] = [:]
  private var closingWorkspaces: Set<String> = []

  func register(
    operationId: String,
    workspaceId: String,
    engineGeneration: Int
  ) -> WorkspaceNativeOperation? {
    cleanupCondition.lock()
    defer { cleanupCondition.unlock() }
    guard !closingWorkspaces.contains(workspaceId) else { return nil }
    guard active[operationId] == nil else { return nil }
    let operation = WorkspaceNativeOperation(workspaceId: workspaceId, engineGeneration: engineGeneration)
    active[operationId] = operation
    return operation
  }

  func remove(_ operationId: String) {
    cleanupCondition.lock()
    active.removeValue(forKey: operationId)
    cleanupCondition.broadcast()
    cleanupCondition.unlock()
  }

  func cancel(_ operationId: String) {
    cleanupCondition.lock()
    let operation = active[operationId]
    cleanupCondition.unlock()
    operation?.cancel()
  }

  /// Prevents new work before requesting cancellation of existing work.
  func closeWorkspace(_ workspaceId: String) {
    cleanupCondition.lock()
    closingWorkspaces.insert(workspaceId)
    let matching = active.values.filter { $0.workspaceId == workspaceId }
    cleanupCondition.unlock()
    matching.forEach { $0.cancel() }
  }

  func isWorkspaceClosing(_ workspaceId: String) -> Bool {
    cleanupCondition.lock()
    defer { cleanupCondition.unlock() }
    return closingWorkspaces.contains(workspaceId)
  }

  func cancelAll() {
    cleanupCondition.lock()
    let values = Array(active.values)
    cleanupCondition.unlock()
    values.forEach { $0.cancel() }
  }

  func hasLiveWorkspace(_ workspaceId: String) -> Bool {
    cleanupCondition.lock()
    defer { cleanupCondition.unlock() }
    return active.values.contains { $0.workspaceId == workspaceId }
  }

  /// Called from a background cleanup owner; it never blocks the UI thread.
  func awaitWorkspaceCleanup(_ workspaceId: String) {
    cleanupCondition.lock()
    while active.values.contains(where: { $0.workspaceId == workspaceId }) {
      cleanupCondition.wait()
    }
    cleanupCondition.unlock()
  }

  var activeCount: Int {
    cleanupCondition.lock()
    defer { cleanupCondition.unlock() }
    return active.count
  }
}
